File: euphony/src/main/java/euphony/lib/transmitter/EuCodeMaker.java
package euphony.lib.transmitter;

import android.util.Log;

import euphony.lib.util.EuOption;
import euphony.lib.util.PacketErrorDetector;

public class EuCodeMaker extends EuFreqGenerator {
	String mMainString;
	short[] mCodeSource;
	private EuOption mTxOption;
	private int mControlPoint;
	private int mOutsetFrequency;
	private int mDataInterval;

	private EuOption.ModulationType mModulation;
	
	public EuCodeMaker() { }

	public EuCodeMaker(EuOption option) {
		mTxOption = option;
		mControlPoint = option.getControlPoint();
		mModulation = option.getModulationType();
		mDataInterval = option.getDataInterval();
		mOutsetFrequency = option.getOutsetFrequency();
	}

	public short[] assembleData(String data)
    {

    	int[] payload = new int[data.length() + 1];
    	short[] assembledData;
    	int payloadSum = 0;
    	
    	switch(mModulation){
			case ASK: {
				assembledData = applyCrossFade(makeStaticFrequency(mOutsetFrequency, 0), CrossfadeType.BOTH);
				for (int i = 0; i < data.length(); i++) {
					char ch = data.charAt(i);
					switch (ch) {
						case '0':
							assembledData = appendRawData(assembledData, getZeroSource());
							break;
						case '1':
							assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint, 0), CrossfadeType.BOTH));
							break;
					}
					payload[i] = ch - '0';
					payloadSum += payload[i];
				}


				int checksum = PacketErrorDetector.makeCheckSum(payloadSum);
				int parity = PacketErrorDetector.makeParellelParity(payload);

				Log.i("euphony_code", "CODE" + data + "CHECKSUM : " + checksum + " PARITY : " + parity);
				assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint + mDataInterval * checksum, 0), CrossfadeType.BOTH));
				assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint  + mDataInterval * parity, 0), CrossfadeType.BOTH));

			}
				break;

			case FSK: {
				assembledData = applyCrossFade(makeStaticFrequency(mOutsetFrequency, 0), CrossfadeType.BOTH);

				for (int i = 0; i < data.length(); i++) {
					char ch = data.charAt(i);
					switch (ch) {
						case '0':
						case '1':
						case '2':
						case '3':
						case '4':
						case '5':
						case '6':
						case '7':
						case '8':
						case '9':
							assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint + mDataInterval * (ch - '0'), 0), CrossfadeType.BOTH));
							payload[i] = ch - '0';
							break;
						case 'a':
						case 'b':
						case 'c':
						case 'd':
						case 'e':
						case 'f':
							assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint + mDataInterval * (ch - 'a' + 10), 0), CrossfadeType.BOTH));
							payload[i] = (ch - 'a') + 10;
							break;
					}
					payloadSum += payload[i];
				}


				int checksum = PacketErrorDetector.makeCheckSum(payloadSum);
				int parity = PacketErrorDetector.makeParellelParity(payload);

				Log.i("euphony_code", "CODE" + data + "CHECKSUM : " + checksum + " PARITY : " + parity);
				assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint + mDataInterval * checksum, 0), CrossfadeType.BOTH));
				assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint  + mDataInterval * parity, 0), CrossfadeType.BOTH));

			}	break;
			default:
			case CPFSK: {
				assembledData = applyCrossFade(makeStaticFrequency(mOutsetFrequency, 0), CrossfadeType.FRONT);
				for (int i = 0; i < data.length(); i++) {
					char ch = data.charAt(i);
					switch (ch) {
						case '0':
						case '1':
						case '2':
						case '3':
						case '4':
						case '5':
						case '6':
						case '7':
						case '8':
						case '9':
							assembledData = appendRawData(assembledData, makeFrequencyWithCP(mControlPoint + mDataInterval * (ch - '0')));
							payload[i] = ch - '0';
							break;
						case 'a':
						case 'b':
						case 'c':
						case 'd':
						case 'e':
						case 'f':
							assembledData = appendRawData(assembledData, makeFrequencyWithCP(mControlPoint + mDataInterval * (ch - 'a' + 10)));
							payload[i] = (ch - 'a') + 10;
							break;
					}
					payloadSum += payload[i];
				}

				int checksum = PacketErrorDetector.makeCheckSum(payloadSum);
				int parity = PacketErrorDetector.makeParellelParity(payload);

				Log.i("euphony_code", "CODE" + data + "CHECKSUM : " + checksum + " PARITY : " + parity);
				assembledData = appendRawData(assembledData, makeFrequencyWithCP(mControlPoint + mDataInterval * checksum));
				assembledData = appendRawData(assembledData, applyCrossFade(makeFrequencyWithCP(mControlPoint  + mDataInterval * parity), CrossfadeType.END));

			}
			break;
		}

		return assembledData;
    }

	public short[] assembleLiveData(String data)
	{
		short[] liveStartData = applyCrossFade(makeStaticFrequency(mOutsetFrequency - mDataInterval, 0), CrossfadeType.BOTH);
		short[] startData = applyCrossFade(makeStaticFrequency(mOutsetFrequency, 0), CrossfadeType.BOTH);
		short[] assembledData = liveStartData.clone();
		int[] payload = new int[data.length() + 1];
		int payloadSum = 0;

		switch(mModulation)
		{
			case ASK:
				for(int i = 0; i < data.length(); i++)
				{
					switch(data.charAt(i))
					{
						case '0':
							assembledData = appendRawData(assembledData, getZeroSource());
							break;
						case '1':
							assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint,0), CrossfadeType.BOTH));
							break;
					}
					// checksum
					assembledData = appendRawData(assembledData, makeFrequencyWithValue(PacketErrorDetector.makeCheckSum(data.charAt(i) - '0')));
					// parity check
					assembledData = appendRawData(assembledData, makeFrequencyWithValue(PacketErrorDetector.makeParellelParity(data.charAt(i) - '0')));
					// final code
					assembledData = appendRawData(assembledData, startData);
				}
				break;

			case FSK:
				for(int i = 0; i < data.length(); i++){
					char ch = data.charAt(i);
					switch(ch)
					{
						case '0':
						case '1':
						case '2':
						case '3':
						case '4':
						case '5':
						case '6':
						case '7':
						case '8':
						case '9':
							assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint + mDataInterval * (ch - '0'), 0), CrossfadeType.BOTH));
							break;
						case 'a':
						case 'b':
						case 'c':
						case 'd':
						case 'e':
						case 'f':
							assembledData = appendRawData(assembledData, applyCrossFade(makeStaticFrequency(mControlPoint + mDataInterval * (ch - 'a' + 10), 0), CrossfadeType.BOTH));
							break;
					}
					// checksum
					assembledData = appendRawData(assembledData, makeFrequencyWithValue(PacketErrorDetector.makeCheckSum(ch - '0')));
					// parity check
					assembledData = appendRawData(assembledData, makeFrequencyWithValue(PacketErrorDetector.makeParellelParity(ch - '0')));
					// final code
					assembledData = appendRawData(assembledData, startData);
				}
				break;
		}

		return assembledData;
	}
}

File: euphony/src/main/kotlin/euphony/lib/transmitter/EuCodeMaker.kt
package euphony.lib.transmitter

import android.util.Log
import euphony.lib.util.PacketErrorDetector

class EuCodeMaker : EuFreqGenerator {
    internal var mMainString: String? = null
    internal var mCodeSource: ShortArray? = null
    var START_BIT = freqBasePoint - freqSpan

    private var mChannelMode = CHANNEL.EXINGLE

    enum class CHANNEL {
        SINGLE, MULTI, EXINGLE
    }

    constructor() {}

    constructor(channelMode: CHANNEL) {
        mChannelMode = channelMode
    }


    fun euAssembleData(data: String): ShortArray {
        var assembledData = euApplyCrossFade(euMakeStaticFrequency(START_BIT, 0))
        val payload = IntArray(data.length + 1)

        when (mChannelMode) {
            EuCodeMaker.CHANNEL.SINGLE -> for (i in 0 until data.length) {
                when (data[i]) {
                    '0' -> assembledData = euAppendRawData(assembledData, zeroSource)
                    '1' -> assembledData = euAppendRawData(assembledData, euApplyCrossFade(euMakeStaticFrequency(freqBasePoint, 0)))
                }
            }

            EuCodeMaker.CHANNEL.MULTI -> for (i in 0 until data.length) {
                when (data[i]) {
                    '0' -> {
                    }
                    '1' -> assembledData = euMixingRawData(assembledData, euApplyCrossFade(euMakeStaticFrequency(freqBasePoint + freqSpan * (i + 1), 0)))
                }
            }

            EuCodeMaker.CHANNEL.EXINGLE -> for (i in 0 until data.length) {
                val ch = data[i]
                when (ch) {
                    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' -> assembledData = euAppendRawData(assembledData, euApplyCrossFade(euMakeStaticFrequency(freqBasePoint + freqSpan * (ch - '0'), 0)))
                    'a', 'b', 'c', 'd', 'e', 'f' -> assembledData = euAppendRawData(assembledData, euApplyCrossFade(euMakeStaticFrequency(freqBasePoint + freqSpan * (ch - 'a' + 10), 0)))
                }
            }
        }

        for (i in 0 until data.length) {
            when (data[i]) {
                '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' -> payload[i] = data[i] - '0'
                'a', 'b', 'c', 'd', 'e', 'f' -> payload[i] = data[i] - 'a' + 10
            }
        }
        val checksum = PacketErrorDetector.makeCheckSum(payload)
        val parity = PacketErrorDetector.makeParellelParity(payload)

        Log.i("UHEHE", "CHECKSUM : $checksum PARITY : $parity")
        assembledData = euAppendRawData(assembledData, euApplyCrossFade(euMakeStaticFrequency(freqBasePoint + freqSpan * checksum, 0)))
        assembledData = euAppendRawData(assembledData, euApplyCrossFade(euMakeStaticFrequency(freqBasePoint + freqSpan * parity, 0)))

        assembledData = euAppendRawData(assembledData, euApplyCrossFade(euMakeStaticFrequency(START_BIT, 0)))

        return assembledData
    }

}